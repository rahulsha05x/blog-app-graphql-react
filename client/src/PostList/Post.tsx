
export interface Post {
    id:string,
    title:string,
    description:string
}

export class Post implements Post {
    constructor(public id:string,public title:string,public description:string) {}
}
