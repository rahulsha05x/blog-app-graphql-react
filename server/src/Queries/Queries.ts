import { gql } from "apollo-boost";

export const LOAD_POSTS = gql`
  query GetPosts {
    posts {
      id
      title
      description
    }
  }
`;
export const DELETE_POSTS = gql`
  mutation DeletePost($id: ID!) {
    deletePost(id: $id) {
      status
    }
  }
`;
export const ADD_POST = gql`
  mutation CreatePost($input: CreatePostInput) {
    post: createPost(input: $input) {
      id
      title
      description
    }
  }
`;

export const UPDATE_POST = gql`
  mutation UpdatePost($id: ID!, $title: String, $description: String) {
    post: updatePost(id: $id, title: $title, description: $description) {
      id
      title
      description
    }
  }
`;
export const GET_POST = gql`
query getPost($id: ID!) {
  post(id: $id) {
    id
    title
    description
  }
}
`