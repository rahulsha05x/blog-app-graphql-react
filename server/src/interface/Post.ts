export interface Post {
  id: string;
  title: string;
  description: string;
}
export interface inputVars {
  id: string;
}
export interface PostResponse {
  success: boolean;
  code: string;
  response?: any;
  error?: any;
}
