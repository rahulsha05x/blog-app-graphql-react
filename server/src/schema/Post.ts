import { Field, Int, ObjectType,ID } from "type-graphql";

@ObjectType()
export class Post {
    @Field(type => ID)
    id:string;
    @Field({nullable:true})
    title:string;
    
    @Field({ nullable: true })
    description:string
}