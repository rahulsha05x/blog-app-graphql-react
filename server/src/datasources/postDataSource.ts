  
import { RESTDataSource } from "apollo-datasource-rest";
import { Post } from "../interface/Post";

export class PostDataSource extends RESTDataSource {
  constructor() {
    super();
    // this.baseURL = "https://api.randomuser.me/";
    this.baseURL = "http://localhost:8080";
  }

  async getPosts() {
    const posts = await this.get("/posts");
    return posts;
  }
  async getPostById(id:string) {
      const post = await this.get(`/posts/${id}`)
      return post;
  }
  async addNewPost({title,description}:Post) {
      const post = await this.post(`/posts`,{title,description});
      return post
  }
  async deletePost(id:string) {
      const post = await this.delete(`/posts/${id}`);
      return {status:"Success"};
  }
  async updatePost({id,title,description}:Post) {
      console.log("id is",id)
      const post = await this.patch(`/posts/${id}`,{id,title,description});
      return post;
  }
}