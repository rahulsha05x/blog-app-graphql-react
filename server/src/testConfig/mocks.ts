import { Post } from "../interface/Post";

export const postsMockResponse = (): Post[] => [
  {
    id: "s-on9Sj",
    title: "New title 13234",
    description: "New description 123",
  },
  {
    id: "WzSLQSz",
    title: "New title 2",
    description: "New description 2",
  },
  {
    id: "ahaELtt",
    title: "New title 31",
    description:
      "The substring() method extracts the characters from a string, between two specified indices, and returns the new sub string.This method extracts the characters in a string between start and end, not including end itself.If start is greater than end, this method will swap the two arguments, meaning str.substring(1, 4) == str.substring(4, 1).If either start or end is less than 0, it is treated as if it were 0.Note: The substring() method does not change the original string.",
  },
  {
    id: "d4e2cSh",
    title: "New title",
    description: "Adding blog post",
  },
];
export const postMockResponse = (): Post => {
  return {
    id: "s-on9Sj",
    title: "New title 13234",
    description: "New description 123",
  };
};
export const postCreateResponseMock = (): Post => {
  return {
    title: "Game of thrones",
    description: "People fighting for the crown",
    id: "yRfeOFz",
  };
};
