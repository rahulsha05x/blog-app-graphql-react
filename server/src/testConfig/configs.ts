import { PostDataSource } from "../datasources/postDataSource";
import { ApolloServer } from "apollo-server";
import { importSchema } from "graphql-import";
import { resolvers } from "../resolver";
import ApolloServerBase,{createTestClient,ApolloServerTestClient} from 'apollo-server-testing';
import {typeDefs} from './schema'



export const constructTestServer = (dataSources:any):ApolloServerTestClient => {
    const postsDataAPI = new PostDataSource();
  
    const server = new ApolloServer({
      typeDefs,
      resolvers,
      dataSources,
    });
  
    return createTestClient(server);
  };