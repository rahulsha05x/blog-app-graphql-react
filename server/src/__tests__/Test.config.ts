import ApolloServerBase,{createTestClient} from 'apollo-server-testing';
import  gql from 'graphql-tag';
import  nock from 'nock';

import {constructTestServer} from '../testConfig/configs'
//const {constructTestServer} = require('./__utils');

// the mocked REST API data
import {postsMockResponse, postMockResponse, postCreateResponseMock} from '../testConfig/mocks';
import { LOAD_POSTS, GET_POST, UPDATE_POST, ADD_POST } from '../Queries';
import { Post } from '../interface/Post';
import { PostDataSource } from '../datasources/postDataSource';
// the mocked SQL DataSource store

const initMockServer = (mockresponse:any) => {
  const postsDataAPI:any = new PostDataSource();
  const getStub = ()=> Promise.resolve(mockresponse());
  postsDataAPI.get = jest.fn(getStub)
  postsDataAPI.patch = jest.fn(getStub)
  postsDataAPI.post = jest.fn(getStub)
  

  // use our test server as input to the createTestClient fn
  // This will give us an interface, similar to apolloClient.query
  // to run queries against our instance of ApolloServer
  const { query,mutate } = constructTestServer(() => ({ postsDataAPI }));
  return {query,postsDataAPI,mutate}
}

describe('Queries', () => {
  beforeAll(()=>{
    
  })
  it("check jest",()=>{
    expect(2).toBe(2);
  })
  it('fetches list of posts', async () => {
    const {query,postsDataAPI} = initMockServer(postsMockResponse);
    const res = await query({query: LOAD_POSTS});
    expect(res.data).toMatchSnapshot();
    expect(res.errors).toBe(undefined);
    expect(postsDataAPI.get).toHaveBeenCalledWith('/posts');
    expect(res?.data?.posts).toEqual(postsMockResponse())
  });
  it('fetch post', async () => {
    const {query,postsDataAPI} = initMockServer(postMockResponse);
    const res = await query({query: GET_POST,variables:{id:'2'}});
    expect(res.data).toMatchSnapshot();
    expect(res.errors).toBe(undefined);
    expect(postsDataAPI.get).toHaveBeenCalledWith('/posts/2');
    expect(res?.data?.post).toEqual(postMockResponse())
  });
  it('mutate post:get Post with ID', async () => {
    const {mutate,postsDataAPI} = initMockServer(postMockResponse);
    const variables = {"description": "ABC Test", "id": "s-on9Sj", "title": "Test ABC"};
    const res = await mutate({mutation: UPDATE_POST,variables});
    expect(res.data).toMatchSnapshot();
    expect(res.errors).toBe(undefined);
    expect(postsDataAPI.patch).toHaveBeenCalledWith('/posts/s-on9Sj',variables);
    expect(res?.data?.post).toEqual(postMockResponse())
  });
  it('mutate post:add New Post', async () => {
    const {mutate,postsDataAPI} = initMockServer(postCreateResponseMock);
    const input = {"description": "ABC Test", "title": "Test ABC"};
    const res = await mutate({mutation: ADD_POST,variables:{input}});
    expect(res.data).toMatchSnapshot();
    expect(res.errors).toBe(undefined);
    expect(postsDataAPI.post).toHaveBeenCalledWith('/posts',input);
    expect(res?.data?.post).toEqual(postCreateResponseMock())
  });
});