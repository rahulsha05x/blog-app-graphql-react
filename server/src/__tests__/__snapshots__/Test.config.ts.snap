// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Queries fetch post 1`] = `
Object {
  "post": Object {
    "description": "New description 123",
    "id": "s-on9Sj",
    "title": "New title 13234",
  },
}
`;

exports[`Queries fetches list of posts 1`] = `
Object {
  "posts": Array [
    Object {
      "description": "New description 123",
      "id": "s-on9Sj",
      "title": "New title 13234",
    },
    Object {
      "description": "New description 2",
      "id": "WzSLQSz",
      "title": "New title 2",
    },
    Object {
      "description": "The substring() method extracts the characters from a string, between two specified indices, and returns the new sub string.This method extracts the characters in a string between start and end, not including end itself.If start is greater than end, this method will swap the two arguments, meaning str.substring(1, 4) == str.substring(4, 1).If either start or end is less than 0, it is treated as if it were 0.Note: The substring() method does not change the original string.",
      "id": "ahaELtt",
      "title": "New title 31",
    },
    Object {
      "description": "Adding blog post",
      "id": "d4e2cSh",
      "title": "New title",
    },
  ],
}
`;

exports[`Queries mutate post:add New Post 1`] = `
Object {
  "post": Object {
    "description": "People fighting for the crown",
    "id": "yRfeOFz",
    "title": "Game of thrones",
  },
}
`;

exports[`Queries mutate post:get Post with ID 1`] = `
Object {
  "post": Object {
    "description": "New description 123",
    "id": "s-on9Sj",
    "title": "New title 13234",
  },
}
`;
